clippy_lints/src/assigning_clones.rs

```diff
@@ -0,0 +1,204 @@
+        if !args.is_empty() {
+            return;
+        }
+
+        let op = if method_name.ident.name == sym::clone {
+            Op::Clone
+        } else if method_name.ident.name == sym!(to_owned) {
+            Op::ToOwned
+        } else {
+            return;
+        };
+
+        if ok_to_suggest(cx, op, assign_target, clone_call).is_some() {
+            suggest(cx, op, assign_expr, assign_target, clone_receiver);
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum Op {
+    Clone,
+    ToOwned,
+}
+
+// Return `Some(())` iff we have confirmed that the call is in fact one we want to replace.
+fn ok_to_suggest<'tcx>(
+    cx: &LateContext<'tcx>,
+    op: Op,
+    assign_target: &hir::Expr<'tcx>,
+    clone_call: &hir::Expr<'tcx>,
+) -> Option<()> {
+    // Check that the call is actually a call to the trait.
+    // TODO: Actually we are currently just checking that the result of the call is
+    // a type that implements the trait, which is a bad proxy for it.
+    let clone_result_type = cx.typeck_results().expr_ty_adjusted(clone_call);
+    if !(implements_trait(cx, clone_result_type, op.expected_trait(cx)?, &[])) {
+        return None;
+    }
+
+    // If we're assigning to a dereferenced reference, then we know the place is already valid.
+    // On the other hand, if the place is a variable or a Box, it might be uninitialized,
+    // in which case the suggestion might be wrong.
+    // TODO: Actually ask whether the place is uninitialized at this point, instead of
+    // guessing based on the syntax and type.
+    let ExprKind::Unary(hir::UnOp::Deref, derefed_target_expr) = assign_target.kind
+        else { return None };
+    if !cx.typeck_results().expr_ty(derefed_target_expr).is_ref() {
+        return None;
+    }
+
+    Some(())
+}
+
+fn suggest<'tcx>(
+    cx: &LateContext<'tcx>,
+    op: Op,
+    assign_expr: &hir::Expr<'tcx>,
+    assign_target: &hir::Expr<'tcx>,
+    clone_receiver: &hir::Expr<'tcx>,
+) {
+    span_lint_and_then(cx, ASSIGNING_CLONES, assign_expr.span, op.message(), |diag| {
+        // TODO: Make this MachineApplicable once we are more certain that the method being called
+        // is what we think it is.
+        let mut applicability = Applicability::MaybeIncorrect;
+
+        diag.span_suggestion(
+            assign_expr.span,
+            op.suggestion_msg(),
+            op.suggested_replacement(cx, assign_target, clone_receiver, &mut applicability),
+            applicability,
+        );
+    });
+}
+
+impl Op {
+    fn expected_trait(self, cx: &LateContext<'_>) -> Option<hir::def_id::DefId> {
+        match self {
+            Op::Clone => cx.tcx.lang_items().clone_trait(),
+            Op::ToOwned => cx.tcx.get_diagnostic_item(sym::ToOwned),
+        }
+    }
+
+    fn message(self) -> &'static str {
+        // TODO: Use the receiver type to say "is" instead of "may be" for types which
+        // are known to have optimizations (e.g. `String`).
```

Comment:
You can use type checking, to get the type of the clone call. If it's an ADT, you can use the `DefId` to check for known types :)

See:
* [Checking for a specific type](https://doc.rust-lang.org/clippy/development/common_tools_writing_lints.html#checking-for-a-specific-type)
* [Diagnostic items](https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-items.html)